File: bindings/node/__tests__/genome.test.js
"use strict";

const { test } = require("node:test");
const assert = require("node:assert");
const { Genome } = require("../index.js");

const SPEC = {
  features: [{ kind: "price", field: "close" }],
  symbols: ["AAA", "BBB", "CCC"],
  normalize: "z_score",
  metric: "euclid",
  seed: 24333,
};

const DATA = {
  AAA: [{ time: 0, open: 1, high: 1, low: 1, close: 1, volume: 0 }],
  BBB: [{ time: 0, open: 2, high: 2, low: 2, close: 2, volume: 0 }],
  CCC: [{ time: 0, open: 100, high: 100, low: 100, close: 100, volume: 0 }],
};

function built() {
  const g = new Genome(JSON.stringify(SPEC));
  const ok = JSON.parse(g.command(JSON.stringify({ cmd: "build", data: DATA })));
  assert.strictEqual(ok.ok, true);
  return g;
}

test("similar excludes self and finds the closest coin", () => {
  const res = JSON.parse(built().command(JSON.stringify({ cmd: "similar", symbol: "AAA", k: 2 })));
  const names = res.neighbors.map((n) => n.symbol);
  assert.ok(!names.includes("AAA"));
  assert.strictEqual(names[0], "BBB");
});

test("anomaly ranks the outlier first", () => {
  const res = JSON.parse(built().command(JSON.stringify({ cmd: "anomaly" })));
  assert.strictEqual(res.anomalies[0].symbol, "CCC");
});

test("deterministic across instances", () => {
  const cmd = JSON.stringify({ cmd: "cluster", k: 2 });
  assert.strictEqual(built().command(cmd), built().command(cmd));
});

test("version", () => {
  assert.strictEqual(typeof new Genome(JSON.stringify(SPEC)).version(), "string");
});

File: bindings/java/src/test/java/org/wickra/genome/GenomeTest.java
package org.wickra.genome;

import static org.junit.jupiter.api.Assertions.assertFalse;
import static org.junit.jupiter.api.Assertions.assertThrows;
import static org.junit.jupiter.api.Assertions.assertTrue;

import org.junit.jupiter.api.Test;

class GenomeTest {
    static final String SPEC =
            "{\"features\":[{\"kind\":\"price\",\"field\":\"close\"}],"
                    + "\"symbols\":[\"AAA\",\"BBB\",\"CCC\"],"
                    + "\"normalize\":\"z_score\",\"metric\":\"euclid\",\"seed\":24333}";

    static final String DATA =
            "{\"AAA\":[{\"time\":0,\"open\":1,\"high\":1,\"low\":1,\"close\":1,\"volume\":0}],"
                    + "\"BBB\":[{\"time\":0,\"open\":2,\"high\":2,\"low\":2,\"close\":2,\"volume\":0}],"
                    + "\"CCC\":[{\"time\":0,\"open\":100,\"high\":100,\"low\":100,\"close\":100,\"volume\":0}]}";

    private static Genome built() {
        Genome g = new Genome(SPEC);
        String ok = g.command("{\"cmd\":\"build\",\"data\":" + DATA + "}");
        assertTrue(ok.contains("\"ok\":true"), ok);
        return g;
    }

    @Test
    void versionIsNonEmpty() {
        assertFalse(Genome.version().isEmpty());
    }

    @Test
    void similarExcludesSelf() {
        try (Genome g = built()) {
            String out = g.command("{\"cmd\":\"similar\",\"symbol\":\"AAA\",\"k\":2}");
            assertTrue(out.startsWith("{\"neighbors\":[{\"symbol\":\"BBB\""), out);
        }
    }

    @Test
    void anomalyRanksOutlierFirst() {
        try (Genome g = built()) {
            String out = g.command("{\"cmd\":\"anomaly\"}");
            assertTrue(out.startsWith("{\"anomalies\":[{\"symbol\":\"CCC\""), out);
        }
    }

    @Test
    void invalidSpecThrows() {
        assertThrows(IllegalArgumentException.class, () -> new Genome("{ not valid json"));
    }
}
